### PR TITLE
[WIP] #261: Add fuzz test

### PR DESCRIPTION
**Pull request by td-bn**
## Overview
WIP PR for [issue#261](https://github.com/TBD54566975/ssi-sdk/issues/261)

## Description
Adds:
- fuzz test for "Create and then resolve returns a document with the same ID".
- Unit tests for "Generate and then resolve returns a document with the same ID".

## How Has This Been Tested?
With commands:
`mage fuzz`
`mage test`

### CODE REVIEW

**Comment by td-bn** [comment]
First time working with `go` 🙏 . With this PR I'm trying to capture expectations and get early feedback.

Looking for pointers to improve and extend this PR - specifically  (but not limited to):
- better way to fuzz test these properties
- changes required to capture such tests in a mage command

File: did/key_fuzz_test.go [hunk 1]
```diff
@@ -0,0 +1,43 @@
+package did
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+)
+
+func FuzzGenerateAndDecode(f *testing.F) {
+	keytypes := GetSupportedDIDKeyTypes()
```
Comment (andresuribe87) [inline did/key_fuzz_test.go:10]: First off, this is awesome; thanks for taking a stab at it!

That said, I realize now that these tests are likely better suited as unit test that go over all keytypes.

Fuzz testing works best when there is a function that can accept arbitrary random input from the fuzz arguments. 

I realize that this isn't the best case for `GenerateDIDKey`. So my suggestion would be to:
* Move the existing tests you wrote to unit tests which go over all keytypes.
* Do similar fuzz tests as what you have, but calling `CreateDIDKey` instead. You can pass in the the `publicKey []byte` as a fuzz argument. For key type, you can do the same thing you're currently doing. 
  * Make sure to verify that the `publicKey` passed in is the same as what results after calling `Decode` :)

**Review by andresuribe87 — COMMENTED** [review]
Regarding mage, I'd recommend adding a separate command for fuzzing, and making sure you include the flag `-fuzztime`. A sensible default can be 10 seconds. 

After you add the mage command, we can include it as part of the CI pipeline. Let me know if that all makes sense and thanks!

File: did/key_fuzz_test.go [hunk 2]
```diff
@@ -6,37 +6,51 @@ import (
...
-	f.Fuzz(func(t *testing.T, a uint) {
-		kt := keytypes[a%uint(len(keytypes))]
 
-		privKey, didKey, err := GenerateDIDKey(kt)
-		assert.NotEmpty(t, privKey)
+	pks := []string{"b9c5714089478a327f09197987f16f9e5d936e8a", "5f246d7d19aa612d6718d27c1da1ee66859586b0", "7d2d43e63666f45b40316b44212325625dbaeb40", "1c1f02f1640e52b313f2d504b3c0c7ee8ad61108", "69c5888ecd21287fbdac5a43d1558bf73c51e38b"}
+	for i, pk := range pks {
+		f.Add(uint(i), []byte(pk))
+	}
+	f.Fuzz(func(t *testing.T, a uint, pubKey []byte) {
```
Comment (decentralgabe) [inline did/key_fuzz_test.go:16]: what is `a` can it be a more meaningful variable name?

File: did/key_fuzz_test.go [hunk 3]
```diff
@@ -6,37 +6,51 @@ import (
...
+func FuzzCreateAndDecode(f *testing.F) {
 	keytypes := GetSupportedDIDKeyTypes()
-	f.Fuzz(func(t *testing.T, a uint) {
-		kt := keytypes[a%uint(len(keytypes))]
 
-		privKey, didKey, err := GenerateDIDKey(kt)
-		assert.NotEmpty(t, privKey)
+	pks := []string{"b9c5714089478a327f09197987f16f9e5d936e8a", "5f246d7d19aa612d6718d27c1da1ee66859586b0", "7d2d43e63666f45b40316b44212325625dbaeb40", "1c1f02f1640e52b313f2d504b3c0c7ee8ad61108", "69c5888ecd21287fbdac5a43d1558bf73c51e38b"}
+	for i, pk := range pks {
+		f.Add(uint(i), []byte(pk))
```
Comment (decentralgabe) [inline did/key_fuzz_test.go:14]: why cast to a uint?

File: did/key_fuzz_test.go [hunk 4]
```diff
@@ -6,37 +6,51 @@ import (
...
+	}
+	f.Fuzz(func(t *testing.T, a uint, pubKey []byte) {
+		kt := keytypes[a%uint(len(keytypes))]
 
+		didKey, err := CreateDIDKey(kt, pubKey)
+		if err != nil {
+			t.Skip()
+		}
 		expectedLdKeyType, _ := KeyTypeToLDKeyType(kt)
```
Comment (decentralgabe) [inline did/key_fuzz_test.go:23]: let's err check here

and nit: `expectedLDKeyType`

File: did/key_test.go [hunk 5]
```diff
@@ -196,6 +196,36 @@ func TestExpand(t *testing.T) {
 	})
 }
 
+func TestGenerateAndDecode(t *testing.T) {
+	for _, kt := range GetSupportedDIDKeyTypes() {
+		privKey, didKey, err := GenerateDIDKey(kt)
+		assert.NotEmpty(t, privKey)
+
+		expectedLdKeyType, _ := KeyTypeToLDKeyType(kt)
```
Comment (decentralgabe) [inline did/key_test.go:204]:
```suggestion
		expectedLLKeyType, _ := KeyTypeToLDKeyType(kt)
```

File: magefile.go [hunk 6]
```diff
@@ -64,6 +68,92 @@ func runTests(extraTestArgs ...string) error {
...
+			if err != nil {
+				return err
+			}
+		}
+
+	}
+	return nil
+}
+
+var FuzzFuncNameRegex = regexp.MustCompile(`\bFuzz\w+`)
```
Comment (decentralgabe) [inline magefile.go:102]: does this need to be exported?

File: magefile.go [hunk 7]
```diff
@@ -64,6 +68,92 @@ func runTests(extraTestArgs ...string) error {
...
+
+// Searches the given directory recursively for Go files containing the word "Fuzz" in their function names.
+// Returns the names of the functions that match the search criteria.
+func getFuzzTests(dir string) ([]string, error) {
+	results := []string{}
```
Comment (decentralgabe) [inline magefile.go:107]:
```suggestion
	var results []string
```

**Review by andresuribe87 — COMMENTED** [review]
Thanks for doing this! One small comment but looks great.

**Comment by andresuribe87** [comment]
Ohhh I forgot - it would make sense to include this in the github CI. Take a look at the `.github/workflows/ci.yml` file.

**Review by andresuribe87 — APPROVED** [review]

**Pull request merged**